### ISSUE TITLE
TypeScript does not catch type that does not match complex union

### ISSUE THREAD

**Issue by jlek** [microsoft/TypeScript#52564]
# Bug Report

### 🔎 Search Terms

Complex union, should not compile

### 🕗 Version & Regression Information

This is the behaviour in every version I tried, and I reviewed the FAQ for entries about unions.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=5.0.0-dev.20230202#code/C4TwDgpgBAwg9gO2AJwJYDcIBMAq5oC8AUFFAD5QDeJptAZnHAFxQBEAQgIIBKr5bXAFqsA3DVIBfGhWq16jFqwCKAVQAao8bQCOAVwAeivfs1yppGVtINmAnnwodOwsXNIAjAIYAvRV++mtBJiRAD0oVA4ABaoAM5QsVFwugA2WFAIcMBQAMZwALZgqCkQADRQ7rrZqNlYcBCxAIREeQix2a0oGNgAovqehSUs8EhomLj4UARUNDaKXLylNMZGBqxLwURAA)

### 💻 Code

```ts
type ContrivedType =
  | {
      foo: "BAR" | "BAZ";
    }
  | {
      foo: "QUX";
      qux: "qux";
    }
  | {
      foo: "BAR" | "BAZ";
      baz: "baz";
    };

// This should not compile, but it does!
const contrivedExample: ContrivedType = {
  foo: "BAR",
  qux: "qux",
};
```

### 🙁 Actual behavior

It compiles.

### 🙂 Expected behavior

It should not compile.

**Comment by MartinJohns** [comment]
Duplicate of #51873. From a type system perspective it's perfectly fine, because your object matches the first type of the union. Objects are not sealed, they can have additional properties. People often don't realize it because of the "excess property checks", which was called more of a linter feature.

**Comment by jlek** [comment]
Thanks for the swift reply @MartinJohns!

> your object matches the first type of the union.

I am not sure if I fully understand what the TypeScript compiler is doing this case. For example, if instead of this union I just try the first case, it catches the error just fine:

```TypeScript
// This does not compile, as expected
const hmmm: { foo: "BAR" | "BAZ" } = {
  foo: "BAR",
  qux: "qux",
};
```

Similarly, it also catches the error if I remove the third object in the union, like so:

```TypeScript
type ContrivedType =
  | {
      foo: "BAR" | "BAZ";
    }
  | {
      foo: "QUX";
      qux: "qux";
    };

// Now this fails to compile, as expected
const contrivedExample: ContrivedType = {
  foo: "BAR",
  qux: "qux"
};
```

I fear I might be misunderstanding something, but this feels inconsistent. :thinking: I'll take some time to read the duplicate issue, maybe it'll help me understand better what's going on. .

**Comment by MartinJohns** [comment]
> For example, if instead of this union I just try the first case, it catches the error just fine:

This is the "excess property check" that I mentioned. This only applies to object literals. This example compiles just fine:

```typescript
const hmmm = { foo: "BAR", qux: "qux" } as const;
const demo: { foo: "BAR" | "BAZ" } = hmmm;
```

> Similarly, it also catches the error if I remove the third object in the union, like so:

Excess property checks does not for work for unions with an overlapping discriminator property. See the issue I linked earlier: #51873. In your first example the discriminator `foo` with the type `"BAR"` overlaps, by removing the third union case you don't have it overlap anymore.

**Comment by jlek** [comment]
Thanks again for the explaination @MartinJohns! Sorry I took to so long to respond, it took me some time to get up to speed with all the concepts like discriminated unions. You're absolutely right of course: this is a duplicate, so I will close this issue.
